File: MARC.Everest/DataTypes/QSP.cs
/* 
 * Copyright 2008-2012 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.

 * 
 * User: Justin Fyfe
 * Date: 09-22-2011
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MARC.Everest.Attributes;
using System.Xml.Serialization;
using MARC.Everest.DataTypes.Interfaces;
using MARC.Everest.Connectors;

namespace MARC.Everest.DataTypes
{
    /// <summary>
    /// Represents a <see cref="T:QSET{T}"/> that has been specialized as a periodic hull between <see cref="P:First"/> and <see cref="P:Second"/>
    /// </summary>
    /// <seealso cref="T:QSET{T}"/>
    /// <seealso cref="T:QSD{T}"/>
    /// <seealso cref="T:QSI{T}"/>
    /// <seealso cref="T:QSP{T}"/>
    /// <seealso cref="T:QSS{T}"/>
    /// <seealso cref="T:QSU{T}"/>
    /// <seealso cref="T:SXPR{T}"/>
    /// <seealso cref="T:SXCM{T}"/>
    /// <seealso cref="T:GTS"/>
    [Serializable]
    [Structure(Name = "QSP", StructureType = StructureAttribute.StructureAttributeType.DataType)]
    [XmlType("QSP", Namespace = "urn:hl7-org:v3")]
    public class QSP<T> : QSET<T>, IEnumerable<ISetComponent<T>>, IEquatable<QSP<T>>
        where T : IAny
    {

        /// <summary>
        /// Creates a new instance of the QSET periodic hull class
        /// </summary>
        public QSP() : base() { }

        /// <summary>
        /// Creates a new instance of the QSET periodic hull class with the specified sets
        /// </summary>
        /// <param name="low">The set used as the basis for the periodic hull operation</param>
        /// <param name="high">The set used for the parameter of the periodic hull function</param>
        public QSP(ISetComponent<T> low, ISetComponent<T> high)
        {
            this.Low = low;
            this.High = high;
        }

        /// <summary>
        /// The set that forms the basis of the hull operation
        /// </summary>
        [Property(Name = "low", Conformance = PropertyAttribute.AttributeConformanceType.Mandatory, PropertyType = PropertyAttribute.AttributeAttributeType.NonStructural)]
        [XmlElement("low")]
        public ISetComponent<T> Low { get; set; }

        /// <summary>
        /// The set used as a parameter to the hull operation
        /// </summary>
        [Property(Name = "high", Conformance = PropertyAttribute.AttributeConformanceType.Mandatory, PropertyType = PropertyAttribute.AttributeAttributeType.NonStructural)]
        [XmlElement("high")]
        public ISetComponent<T> High { get; set; }

        /// <summary>
        /// Validate that the QSET is valid. A QSET is valid when it contains <see cref="P:First"/> and
        /// <paramref name="P:Second"/> and contains no-null terms.
        /// </summary>
        public override bool Validate()
        {
            bool isValid = (this.NullFlavor != null) ^ (this.Low != null && !this.Low.IsNull && this.High != null && !this.High.IsNull);
            return isValid;
        }

        /// <summary>
        /// Validate that the QSET is valid. A QSET is valid when it contains <see cref="P:First"/> and
        /// <paramref name="P:Second"/> and contains no-null terms.
        /// </summary>
        public override IEnumerable<Connectors.IResultDetail> ValidateEx()
        {
            List<IResultDetail> retVal = new List<IResultDetail>();
            if(!((this.NullFlavor != null) ^ (this.Low != null && !this.Low.IsNull && this.High != null && !this.High.IsNull)))
                retVal.Add(new DatatypeValidationResultDetail(ResultDetailType.Error, "QSP", ValidationMessages.MSG_NULLFLAVOR_WITH_VALUE, null));
            if (this.Low == null || this.Low.IsNull)
                retVal.Add(new DatatypeValidationResultDetail(ResultDetailType.Error, "QSP", String.Format(ValidationMessages.MSG_DEPENDENT_VALUE_MISSING, "Low", String.Format("ISetComponent<{0}>", typeof(T).Name)), null));
            if (this.High == null || this.High.IsNull)
                retVal.Add(new DatatypeValidationResultDetail(ResultDetailType.Error, "QSP", String.Format(ValidationMessages.MSG_DEPENDENT_VALUE_MISSING, "High", String.Format("ISetComponent<{0}>", typeof(T).Name)), null));
            return retVal;
        }

        /// <summary>
        /// Periodic hull
        /// </summary>
        protected override SetOperator GetEquivalentSetOperator()
        {
            return SetOperator.PeriodicHull;
        }

        #region IEnumerable<ISetComponent<T>> Members

        /// <summary>
        /// Get the enumerator
        /// </summary>
        public IEnumerator<ISetComponent<T>> GetEnumerator()
        {
            return new List<ISetComponent<T>> { this.Low, this.High }.GetEnumerator();
        }

        #endregion

        #region IEnumerable Members

        /// <summary>
        /// Get the enumerator
        /// </summary>
        /// <returns></returns>
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        #endregion
        
        #region IEquatable<QSP<T>> Members

        /// <summary>
        /// Determine equality between two QSS instances
        /// </summary>
        public bool Equals(QSP<T> other)
        {
            bool result = base.Equals(other);
            result &= this.High == null ? other.High == null : this.High.Equals(other.High);
            result &= this.Low == null ? other.Low == null : this.Low.Equals(other.Low);
            return result;
        }

        /// <summary>
        /// Determine equality between two items
        /// </summary>
        public override bool Equals(object obj)
        {
            if (obj is QSP<T>)
                return Equals((QSP<T>)obj);
            return base.Equals(obj);
        }

        #endregion    

        /// <summary>
        /// Normalize
        /// </summary>
        public override MARC.Everest.Interfaces.IGraphable Normalize()
        {
            QSP<T> retVal = this.Clone() as QSP<T>;
            if (retVal.Low is SXPR<T>)
                retVal.Low = (retVal.Low as SXPR<T>).TranslateToQSET();
            if (retVal.High is SXPR<T>)
                retVal.High = (retVal.High as SXPR<T>).TranslateToQSET();
            return retVal;
        }
    }
}

File: jEverest/src/ca/marc/everest/datatypes/generic/QSP.java
/* 
 * Copyright 2008-2012 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * User: Justin Fyfe
 * Date: 10-11-2012
 */
package ca.marc.everest.datatypes.generic;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import ca.marc.everest.annotations.*;
import ca.marc.everest.datatypes.EverestValidationMessages;
import ca.marc.everest.datatypes.SetOperator;
import ca.marc.everest.datatypes.interfaces.IAny;
import ca.marc.everest.datatypes.interfaces.ISetComponent;
import ca.marc.everest.interfaces.IGraphable;
import ca.marc.everest.interfaces.IResultDetail;
import ca.marc.everest.interfaces.ResultDetailType;
import ca.marc.everest.resultdetails.DatatypeValidationResultDetail;

/**
 * Represents a QSET{T} that has been specialized as a difference between two other 
 * expressions
 */
@Structure(name = "QSP", structureType = StructureType.DATATYPE)
public class QSP<T extends IAny> extends QSET<T> implements Iterable<ISetComponent<T>> {

	// The minuend of this difference
	private ISetComponent<T> m_low;
	// The subtrahend of this difference
	private ISetComponent<T> m_high;
	
	/**
	 * Creates a new instance of the QSP{T} class
	 */
	public QSP() { super(); }
	/**
	 * Creates a new instance of the QSP class with the specified low and high
	 * @param low The initial value of the low
	 * @param high The initial value of the high
	 */
	public QSP(ISetComponent<T> low, ISetComponent<T> high)
	{
		this.m_low = low;
		this.m_high = high;
	}
	
	/**
	 * Gets the value representing the low of the periodic hull expression
	 */
	@Property(name = "low", propertyType = PropertyType.NONSTRUCTURAL, conformance = ConformanceType.MANDATORY)
	public ISetComponent<T> getLow() {
		return this.m_low;
	}
	/**
	 * Sets the value representing the low of the hull expression
	 */
	public void setLow(ISetComponent<T> value) {
		this.m_low = value;
	}
	/**
	 * Gets the value representing the hull expression
	 */
	@Property(name = "high", propertyType = PropertyType.NONSTRUCTURAL, conformance = ConformanceType.MANDATORY)
	public ISetComponent<T> getHigh() {
		return this.m_high;
	}
	/**
	 * Sets a value representing the hull expression
	 */
	public void setHigh(ISetComponent<T> value) {
		this.m_high = value;
	}
	/**
	 * Normalize the expression
	 */
	@SuppressWarnings("unchecked")
	@Override
	public IGraphable normalize() {
		QSP<T> retVal = (QSP<T>)this.shallowCopy();
        if (retVal.m_low instanceof SXPR<?>)
            retVal.m_low = ((SXPR<T>)retVal.m_low).translateToQSET();
        if (retVal.m_high instanceof SXPR<?>)
            retVal.m_high = ((SXPR<T>)retVal.m_high).translateToQSET();
        return retVal;
	}

	/**
	 * Get the iterator
	 */
	@Override
	public Iterator<ISetComponent<T>> iterator() {
		ArrayList<ISetComponent<T>> iterable = new ArrayList<ISetComponent<T>>();
		iterable.add(this.m_low);
		iterable.add(this.m_high);
		return iterable.iterator();
	}

	/**
	 * Get the equivalent set operator
	 */
	@Override
	protected SetOperator getEquivalentSetOperator() {
		return SetOperator.PeriodicHull;
	}
	/** (non-Javadoc)
	 * @see ca.marc.everest.datatypes.ANY#validateEx()
	 */
	@Override
	public Collection<IResultDetail> validateEx() {
		List<IResultDetail> retVal = new ArrayList<IResultDetail>();
        if (!(this.isNull() ^ (this.m_low != null && !this.m_low.isNull() && this.m_high != null && !this.m_high.isNull())))
            retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "QSP", EverestValidationMessages.MSG_NULLFLAVOR_WITH_VALUE, null));
        if (this.m_low == null || this.m_low.isNull())
            retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "QSP", String.format(EverestValidationMessages.MSG_DEPENDENT_VALUE_MISSING, "Minuend", String.format("ISetComponent<{0}>", "?")), null));
        if (this.m_high == null || this.m_high.isNull())
            retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "QSP", String.format(EverestValidationMessages.MSG_DEPENDENT_VALUE_MISSING, "Subtrahend", String.format("ISetComponent<{0}>", "?")), null));
        return retVal;
	}
	/** (non-Javadoc)
	 * @see ca.marc.everest.datatypes.HXIT#validate()
	 */
	@Override
	public boolean validate() {
		boolean isValid = this.isNull() ^ (this.m_low != null && !this.m_low.isNull() && this.m_high != null && !this.m_high.isNull());
        return isValid;
	}
	/** (non-Javadoc)
	 * @see java.lang.Object#hashCode()
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result
				+ ((m_low == null) ? 0 : m_low.hashCode());
		result = prime * result
				+ ((m_high == null) ? 0 : m_high.hashCode());
		return result;
	}
	/** (non-Javadoc)
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	@SuppressWarnings("unchecked")
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		if (getClass() != obj.getClass())
			return false;
		QSP<T> other = (QSP<T>) obj;
		if (m_low == null) {
			if (other.m_low != null)
				return false;
		} else if (!m_low.equals(other.m_low))
			return false;
		if (m_high == null) {
			if (other.m_high != null)
				return false;
		} else if (!m_high.equals(other.m_high))
			return false;
		return true;
	}
	

}
